index.bs

```diff
@@ -475,11 +475,47 @@ When the <dfn method for="XRSession">requestReferenceSpace(|options|)</dfn> meth
 
 </div>
 
+Each {{XRSession}} has a <dfn>list of active XR input sources</dfn> (a [=/list=] of {{XRInputSource}}) which MUST be initially an empty [=/list=].
+
 <div class="algorithm" data-algorithm="get-input-sources">
 
 When the <dfn method for="XRSession">getInputSources()</dfn> method is invoked, the user agent MUST run the following steps:
 
-  1. Return the current [=list of active input sources=].
+  1. Return the current [=list of active XR input sources=].
+
+</div>
+
+The user agent MUST monitor any [=XR input source=]s associated with the [=/XR Device=], including detecting when [=XR input source=]s are added, removed, or changed.
+
+<div class="algorithm" data-algorithm="on-input-source-added">
+
+When a new [=XR input source=] becomes available for {{XRSession}} |session|, the user agent MUST run the following steps:
+
+  1. Let |inputSource| be a new {{XRInputSource}}.
+  1. Add |inputSource| to |session|'s [=list of active XR input sources=] associated with the new [=XR input source=].
+  1. [=Queue a task=] that fires a {{XRSessionEvent}} named {{inputsourceschange!!event}} on |session|.
+
+</div>
+
+<div class="algorithm" data-algorithm="on-input-source-removed">
+
+When an [=XR input source=] is no longer available for {{XRSession}} |session|, the user agent MUST run the following steps:
+
+  1. Let |inputSource| be the {{XRInputSource}} in |session|'s [=list of active XR input sources=] associated with the [=XR input source=] that is not longer available.
+  1. Remove |inputSource| from |session|'s [=list of active XR input sources=].
+  1. [=Queue a task=] that fires a {{XRSessionEvent}} named {{inputsourceschange!!event}} on |session|.
+
+</div>
+
+<div class="algorithm" data-algorithm="on-input-source-change">
+
+When the intrinsic properties of an [=XR input source=] change for {{XRSession}} |session|, such as the {{XRInputSource/handedness}}, {{XRInputSource/targetRayMode}}, or presence of a {{XRInputSource/gripSpace}} or {{XRInputSource/gamepad}}, the user agent MUST run the following steps:
+
+  1. Let |oldInputSource| be the {{XRInputSource}} in |session|'s [=list of active XR input sources=] previously associated with the [=XR input source=] that changed.
+  1. Let |newInputSource| be a new {{XRInputSource}}.
+  1. Remove |oldInputSource| from |session|'s [=list of active XR input sources=].
```

Comment:
Should we also unset an `active` flag on the source? Make `getPose()` on inactive input source spaces returns null (or throw an error for better diagnostics?).

In other words, force people to reload input source lists whenever they change. Otherwise they'll continue using a nonexistant input space.

Reply:
Good suggestion! I'll incorporate that.

Reply:
Following up: This is ever-so-slightly more complex than I had originally anticipated, primarily because I think the "active" flag should actually reside on the space and not the input source. That implies that we probably want to make more consistent use of it elsewhere in the spec, so I wanted to give it some more thought before jumping into a change like that.

Reply:
Totally okay with it being a followup, just something to keep in mind.